CI: Verifier tests: Keep generated object files and logs on test failure

qmonnet opened the pull request:
The CI workflow for testing Cilium's programs against the verifier is supposed to upload its log files and the generated files on failures. However, we lose some of this data, because:

- Multiple jobs (for different kernel versions) upload their log and object files to the same artifact, overwriting previous uploads in the process, so that we only keep the files uploaded by the last job to complete. We can address that by using different artifact names for the different jobs.

- The test itself runs `make -C bpf/ clean` before testing the programs of each source file, thus removing the object files previously generated. Instead, we can clean up the directory just once before running the tests, not between two tests.

- At last, each source file is compiled multiple times with different sets of options, resulting in only the last version of this object file being kept. The solution implemented here consists in renaming these object files (on test failure) to make sure we keep them around.

qmonnet committed:
ci/datapath-verifier: Don't overwrite artifacts between jobs

In the tests-datapath-verifier workflow, all jobs generated from the
matrix (for running the tests with different kernel versions) upload
files to the same artifact on failure. This means that a job may
overwrite the upload from a previous job, and in the end we only get the
uploads from the last job to complete. If several jobs fail with
different errors, then we lose useful debugging information.

Let's have the jobs upload their files to separate artifacts, named
after the kernel version in use.

Signed-off-by: Quentin Monnet <quentin@isovalent.com>

qmonnet committed:
test/verifier: Avoid pruning object files before testing the next file

The verifier tests include a run of "make -C bpf/ clean" prior to
testing the programs in each BPF source files. While this sounds like a
sane practice, this means that at the end of the run, we're left with
only one object file under bpf/, the one used by the last tests
(namely: bpf_sock.o). While the object files are trivial to rebuild
locally, this prevents the CI workflow to find and upload the relevant
object files for instances where the verifier's logs are not enough to
debug.

Let's run instead a single "make -C bpf/ clean" on startup.

Signed-off-by: Quentin Monnet <quentin@isovalent.com>

tklauser commented on test/verifier/verifier_test.go:
```diff
@@ -179,6 +181,10 @@ func TestVerifier(t *testing.T) {
 				if errors.As(err, &ve) {
+					// Rename object file to avoid subsequent runs to overwrite it,
+					// so we can keep it for CI's artifact upload.
+					os.Rename(objFile, path.Join(*ciliumBasePath, "bpf", fmt.Sprintf("%s_%d.o", bpfProgram.name, i)))
```
nit: this could reuse `name` here which is defined here: https://github.com/cilium/cilium/pull/25862/files#diff-941fe2ebb6a483a4cc321ef3fff0f95f576b45f810bce9416eaefb35b46587b7R133

```suggestion
					os.Rename(objFile, path.Join(*ciliumBasePath, "bpf", name + ".o"))
```

qmonnet replied:
Makes sense, I'll update. Thanks!

tklauser reviewed (approved):
Nice, thanks! One nit inline in case a respin is needed.

viktor-kurchenko commented on test/verifier/verifier_test.go:
```diff
@@ -179,6 +181,10 @@ func TestVerifier(t *testing.T) {
 				if errors.As(err, &ve) {
+					// Rename object file to avoid subsequent runs to overwrite it,
+					// so we can keep it for CI's artifact upload.
+					os.Rename(objFile, path.Join(*ciliumBasePath, "bpf", name+".o"))
```
Ignored error, maybe log it at least for better visibility?

viktor-kurchenko reviewed (approved):
LGTM.

julianwiedmann reviewed (approved):
lgtm

qmonnet committed:
test/verifier: Preserve generated object files over multiple runs

For a given source file, we build the object multiple times, with a
different set of build options. This results in object files being lost
on failure. To address this, we rename the object files to include the
iteration number, so that they're still present when we upload to the CI
artifact.

Signed-off-by: Quentin Monnet <quentin@isovalent.com>
